### PR TITLE
Auto aim

### PR DESCRIPTION
**Pull request by stoopipoopy**

### CODE REVIEW

File: src/main/java/frc/robot/FieldConstants.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+/** Add your docs here. */
+public class FieldConstants {
+
+  public static final Translation2d BLUE_SPEAKER_POSE = new Translation2d(-0.086473, 5.757474);
+  public static final Translation2d RED_SPEAKER_POSE = new Translation2d(16.389722, 5.757474);
+
+  public static Translation2d getSpeaker() {
+    // Logger.recordOutput("red speaker", RED_SPEAKER_POSE);
+    // Logger.recordOutput("blue speaker", BLUE_SPEAKER_POSE);
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/FieldConstants.java:19]: remove commented out code

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -93,16 +94,18 @@ public void robotInit() {
...
-            Commands.sequence(
-                Commands.print("fender shot"), Commands.waitSeconds(1.0), Commands.print("pew!")),
-            swerve.stopCmd()));
-    NamedCommands.registerCommand("intake", Commands.print("intake"));
+                    -controller.getRightX() * SwerveSubsystem.MAX_ANGULAR_SPEED)));
+
+    // Controller bindings here
+    controller.start().onTrue(Commands.runOnce(() -> swerve.setYaw(Rotation2d.fromDegrees(0))));
+
+    controller
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/Robot.java:102]: add a comment marking this as a test binding

File: src/main/java/frc/robot/subsystems/autoaim/AutoAim.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.autoaim;
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/autoaim/AutoAim.java:5]: autoaim shouldnt be in subsystems, maybe put it in utils

File: src/main/java/frc/robot/subsystems/autoaim/InterpolatingShotTree.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.TreeMap;
+
+/** Add your docs here. */
+public class InterpolatingShotTree {
+
+  private final TreeMap<Double, ShotData> map = new TreeMap<>();
+
+  public InterpolatingShotTree() {
+
+    // TODO Auto-generated constructor stub
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/autoaim/InterpolatingShotTree.java:16]: clean this up

File: src/main/java/frc/robot/subsystems/autoaim/InterpolatingShotTree.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.autoaim;
+
+import java.util.TreeMap;
+
+/** Add your docs here. */
+public class InterpolatingShotTree {
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/autoaim/InterpolatingShotTree.java:10]: add a doc comment

File: src/main/java/frc/robot/subsystems/autoaim/ShotData.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.autoaim;
+
+import edu.wpi.first.math.interpolation.Interpolatable;
+
+public class ShotData implements Interpolatable<ShotData> {
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/autoaim/ShotData.java:9]: add a doc comment

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 7]
```diff
@@ -150,6 +174,21 @@ public void periodic() {
       Logger.recordOutput("SwerveStates/SetpointsOptimized", new SwerveModuleState[] {});
     }
 
+    double distance =
+        Math.sqrt(
+            Math.pow((pose.getX() - FieldConstants.getSpeaker().getX()), 2)
+                + Math.pow((pose.getY() - FieldConstants.getSpeaker().getY()), 2));
+    ShotData shotData = AutoAim.shotMap.get(distance);
+    Logger.recordOutput("Distance", distance);
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:182]: clean up these debugging logs

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 8]
```diff
@@ -181,12 +220,15 @@ public void periodic() {
     }
   }
 
-  private void runVelocity(ChassisSpeeds speeds) {
+  public void runVelocity(ChassisSpeeds speeds) {
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:223]: this should be private

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 9]
```diff
@@ -299,4 +357,40 @@ public static Translation2d[] getModuleTranslations() {
...
+                  y.getAsDouble(),
+                  calculated + headingController.getSetpoint().velocity);
+            })
+        .beforeStarting(
+            () ->
+                headingController.reset(
+                    new State(
+                        getPose().getRotation().getRadians(),
+                        getVelocity().omegaRadiansPerSecond)));
+    // return this.runVelocityFieldRelative(() -> new ChassisSpeeds(1,1,1));
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:394]: remove

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 10]
```diff
@@ -299,4 +357,40 @@ public static Translation2d[] getModuleTranslations() {
...
   }
+
+  public Rotation2d getRotationToTranslation(Translation2d translation) {
+
+    double angle = Math.atan2(translation.getY() - pose.getY(), translation.getX() - pose.getX());
+    return Rotation2d.fromRadians(angle);
+  }
+
+  public Command pointTowardsTranslation(DoubleSupplier x, DoubleSupplier y) {
+    ProfiledPIDController headingController =
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:368]: make sure having this as a member doesnt break it if you have multiple pointTowardsTranslation calls or call the method twice

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 11]
```diff
@@ -299,4 +357,40 @@ public static Translation2d[] getModuleTranslations() {
...
+        new ProfiledPIDController(1.0, 0.0, 0.0, new Constraints(10, 10));
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:369]: switch constraints to use constants

File: src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java [hunk 12]
```diff
@@ -299,4 +357,40 @@ public static Translation2d[] getModuleTranslations() {
...
+            () -> {
+              Logger.recordOutput(
+                  "PIDController Setpoint", headingController.getSetpoint().position);
+              Logger.recordOutput("PIDController Error", headingController.getPositionError());
+              double calculated =
+                  headingController.calculate(
+                      getPose().getRotation().getRadians(),
+                      getRotationToTranslation(FieldConstants.getSpeaker()).getRadians());
+
+              Logger.recordOutput("PIDController calculate", calculated);
```
Comment (Lewis-Seiden) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java:382]: remove log calls if not needed or organize under folder

**Review by Lewis-Seiden — APPROVED** [review]

File: src/main/java/frc/robot/utils/autoaim/AutoAim.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+package frc.robot.utils.autoaim;
+
+/** Add your docs here. */
+public class AutoAim {
+
+  public static final InterpolatingShotTree shotMap = new InterpolatingShotTree();
+
+  public AutoAim() {
+
+    // Examples ultil we aquire actual data
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/utils/autoaim/AutoAim.java:14]: grammer

File: src/main/java/frc/robot/utils/autoaim/InterpolatingShotTree.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+/*
+ * Class for a TreeMap of ShotData that interpolates between unknown values
+ * Modified version of WPILib's InterpolatingTreeMap
+ */
+public class InterpolatingShotTree {
+
+  private final TreeMap<Double, ShotData> map = new TreeMap<>();
+
+  public InterpolatingShotTree() {}
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/utils/autoaim/InterpolatingShotTree.java:17]: delete whitespace, like `{}`

File: src/main/java/frc/robot/utils/autoaim/ShotData.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.utils.autoaim;
+
+import edu.wpi.first.math.interpolation.Interpolatable;
+
+/*
+ * Holds data about each shot in which we interpolate with
+ */
+public class ShotData implements Interpolatable<ShotData> {
```
Comment (Awesomeplayer165) [inline src/main/java/frc/robot/utils/autoaim/ShotData.java:12]: add line break between comments

**Review by Awesomeplayer165 — COMMENTED** [review]
minor changes

**Comment by stoopipoopy** [comment]
you're

**Review by Lewis-Seiden — APPROVED** [review]
yroureour

**Comment by stoopipoopy** [comment]
y'oture

**Review by Awesomeplayer165 — APPROVED** [review]
muyyummuy

**Pull request merged**